KAFKA-14196: Pausing partition to prevent duplication when autocommit is enabled

philipnee opened the pull request:
**Problem**
Several flaky tests under OffsetValidationTest are indicating an underlying consumer duplication issue.  This is caused by [KAFKA-14024](https://issues.apache.org/jira/browse/KAFKA-14024) after switching the autocommit to async, and therefore causes the fetcher to move ahead without committing the previous offsets. The progress will be lost after partition revocation

**Fix**
Add an internal flag to stop the fetcher from sending fetches, for the partition to be revoked.  The reason we didn't use _pause()_ is because we don't want user API to interfere with the behavior, which might potentially introduce more bugs.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

philipnee committed:
Pausing partition to prevent duplication

philipnee committed:
move pause to autocommit

philipnee committed:
refactor, add cooperative strategy for testing.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -756,10 +756,15 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
             joinPrepareTimer.update();
         }
 
+        final SortedSet<TopicPartition> partitionsToRevoke = getPartitionsToRevoke(protocol, generation, memberId);
```
Do we need to recompute the partitions right before the revocation?

hachikuji replied:
Is your concern that the subscription could change in between the time that we  pause the partitions and the time that the revocation callback is triggered?

philipnee replied:
Right, because if the consumer is assigned to a new partition, we will need to revoke them for eager protocol. (one case i can think of).

hachikuji replied:
I am not sure how we could be assigned new partitions before completing revocation. Is that possible? 

Another case from `getPartitionsToRevoke` below is when the subscription changes in the cooperative protocol. We use the current subscription to filter the partitions that can be revoked immediately before rejoining the group. But for partitions which are no longer subscribed, there is no need to send an offset commit and there is no danger of fetches returning. So I'm wondering if that case can be handled separately. In other words, can we delay computation of this set for the cooperative protocol until we are ready to call `onPartitionsRevoked`?

philipnee replied:
i think it makes sense.  We probably can compute the partition during the revocation due to the nature of cooperative protocol.

philipnee committed:
add cooperative sticky assignor to test

philipnee committed:
cooperative sticky assignor should be tested

philipnee committed:
Merge branch 'trunk' into pause_partition_to_prevent_duplication

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -799,64 +804,95 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
+        //  revocation.
+
+        log.debug("Pausing partitions {} before onJoinPrepare", partitions);
+        partitions.forEach(tp -> subscriptionState().pause(tp));
```
I have mixed feelings about reusing the pause mechanism here. On the one hand, it does what we want. On the other hand, the pause state can be mutated by the user. What if the user calls `resume()` on a partition that we paused internally? Sounds crazy perhaps, but I think I'd rather have a mechanism that can only be accessed internally for stuff like this.

philipnee replied:
Good call there, I think @guozhangwang originally propose marking these partitions using a different flag, to achieve what pause does.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -756,10 +756,15 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
-        if (autoCommitEnabled && autoCommitOffsetRequestFuture == null) {
-            autoCommitOffsetRequestFuture = maybeAutoCommitOffsetsAsync();
+        if (autoCommitEnabled) {
+            pausePartitions(partitionsToRevoke);
```
It seems like this bug (and some of the complexity in this patch) is due to the fact that we do the auto-commit prior to revoking partitions. I wonder if that is really necessary. If we revoke first, then the partitions would be removed from `SubscriptionState` and we wouldn't have to worry about fetches for these partitions returning. Could that work as well?

philipnee replied:
but if we don't commit these offsets, i wonder if that breaks the poll contract, i.e. we need to commit the acked data.  Note that, I believe some of the data could already be returned to the client from the previous poll loop.

showuon replied:
I think what Jason suggested, is change from 
"committing offset for subscribed partitions -> revoking", 
to "revoking -> committing offset for subscribed partitions **+ revoked partitions**". 

I don't think it is a good idea because that would open a door to allow consumer committing offsets for partitions it doesn't own.

hachikuji reviewed (commented):
Thanks, left a few questions.

philipnee committed:
make partition unfetchable

philipnee committed:
Clean up according to PR comments.

clean up

clean up

clean up

clean up

nit

delay the cooperative partition revocation

PR comment: Use switch statement

modify test

PR comments: only precompute EAGER partitions

PR comments

clean up

Reverting to the previous form to simply patching

clean up

rename function

philipnee commented:
Thanks @showuon @hachikuji for the inputs, I just made this PR reviewable, so please review it if possible :)

dajac commented:
@philipnee Thanks for the PR. Could you please update the description to explain the change?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -799,64 +804,94 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
+        //  This only applies to autocommits, as we expect user to handle the offsets menually during the partition
+        //  revocation.
+        log.debug("Marking assigned partitions unconsumable: {}", partitions);
+        partitions.forEach(subscriptions::markUnconsumable);
```
nit: how about we make `SubscriptionState.markUnconsumable` batched? Then we just need to grab the lock once.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -769,6 +773,7 @@ private static class TopicPartitionState {
         private Long logStartOffset; // the log start offset
         private Long lastStableOffset;
         private boolean paused;  // whether this partition has been paused by the user
+        private boolean consumable;
```
How about calling this `pendingRevocation` or something like that? That might make the usage clearer.

showuon replied:
I like the name: `pendingRevocation`, too.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -799,64 +804,95 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
+
+        // only revoke those partitions that are not in the subscription any more.
+        if (protocol == RebalanceProtocol.COOPERATIVE) {
+            // Delay the partition revocation because we don't revoke the already owned partitions
```
leaving this purely for documentation purpose.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -799,64 +804,95 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
+            return partitions;
+        }
+
+        log.warn("Invalid protocol: {}. No partition will be revoked.", protocol);
```
I think this code is dead? Why don't we use a `switch` like we had before? Then the compiler can help us ensure we handle new cases.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -799,64 +804,95 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
-        }
-        return true;
+    private void markPartitionsUnconsumable(final Set<TopicPartition> partitions) {
+        // KAFKA-14196 for more detail, we pause the partition from consumption to prevent duplicated
```
nit: We can leave off the jira reference if we provide enough documentation to understand the issue. I would say something like this:
> When asynchronously committing offsets prior to the revocation of a set of partitions, there will be a window of time between when the offset commit is sent and when it returns and revocation completes. It is possible for pending fetches for these partitions to return during this time, which means the application's position may get ahead of the committed position prior to revocation. This can cause duplicate consumption. To prevent this, we mark the partitions as "pending revocation," which stops the `Fetcher` from sending new fetches or returning data from previous fetches to the user.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -799,64 +804,95 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
-            throw new KafkaException("User rebalance callback throws an error", exception);
-        }
-        return true;
+    private void markPartitionsUnconsumable(final Set<TopicPartition> partitions) {
```
nit: update name

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -799,64 +804,95 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
-
-                    if (!revokedPartitions.isEmpty()) {
-                        exception = invokePartitionsRevoked(revokedPartitions);
+                    partitions.addAll(ownedPartitions.stream()
```
Perhaps we could put this in a helper called `revokeUnsubscribedPartitions` and add a comment:
> For the cooperative strategy, partitions are usually revoked in `onJoinComplete` when the coordinator explicitly requests revocation in order to assign them to another group member. However, any partitions of topics which are no long subscribed can be revoked immediately before rejoining the group.

If we do that, perhaps there's no need to mutate `partitions`. We're already building a new collection here, so we can just make it a new SortedSet.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -799,64 +804,92 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
+        SortedSet<TopicPartition> partitions = new TreeSet<>(COMPARATOR);
+        if (generation == Generation.NO_GENERATION.generationId ||
+                memberId.equals(Generation.NO_GENERATION.memberId)) {
+            partitions.addAll(subscriptions.assignedPartitions());
```
Hmm, these are lost partitions, not ones that we will revoke. We don't need to pause them since any offset commits will get rejected anyway. I feel like we could handle this case in `revokePartitions` as in the old logic. Then we just have the "eager" partitions that are considered here and we could rename this to `eagerPartitionsToRevoke` or something like that.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -799,64 +804,92 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
+                break;
+
+            case COOPERATIVE:
+                // Delay the partition revocation because we don't revoke the already owned partitions
```
I was looking into the cooperative code path. We revoke the partitions in `onJoinComplete`, so that made me wonder why we don't have the same issue there. In fact, there is no additional offset commit in the current logic, which makes me think that the cooperative logic would already be more prone to duplicate consumption. We don't need to fix this here since it seems to be a pre-existing issue, but I am wondering if the failing system tests also cover cooperative assignment?

philipnee replied:
I don't think we have system test covering the cooperative strategy, so it is quite possible that the duplication here.  I think it would be good to update these tests with cooperative strategy.

showuon replied:
Agree that we handle the cooperative issue separately.

hachikuji replied:
I filed this issue: https://issues.apache.org/jira/browse/KAFKA-14224.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -859,36 +848,49 @@ private Optional<Exception> revokePartitions(SortedSet<TopicPartition> partition
-        //  revocation.
-        log.debug("Marking assigned partitions unconsumable: {}", partitions);
+    private Optional<Exception> revokeUnsubscribedPartitions() {
+        //For the cooperative strategy, partitions are usually revoked in onJoinComplete when the
```
nit: space after `//`

Can we move this comment into the `COOPERATIVE` case in `revokePartitions`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -756,12 +756,11 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
 
-        final SortedSet<TopicPartition> partitionsToRevoke = getPartitionsToRevoke(protocol, generation, memberId);
-
+        final SortedSet<TopicPartition> eagerPartitionsToRevoke = eagerPartitionsToRevoke(protocol);
```
Hmm, thinking about this a little more since we're down to just the eager protocol. Since the assignment won't change until the rebalance completes, maybe we do not need to precompute it. In other words, maybe we can restore the original logic in `revokePartitions` and we can change `markPendingPartitions` to something like this:

```java
private void maybeMarkPartitionsPendingRevocation() {
  if (protocol == RebalanceProtocol.EAGER) {
    // When asynchronously committing offsets prior to the revocation of a set of partitions, there will be a
    // window of time between when the offset commit is sent and when it returns and revocation completes. It is
    // possible for pending fetches for these partitions to return during this time, which means the application's
    // position may get ahead of the committed position prior to revocation. This can cause duplicate consumption.
    // To prevent this, we mark the partitions as "pending revocation," which stops the Fetcher from sending new
    // fetches or returning data from previous fetches to the user.
    Set<TopicPartition> partitions = subscriptions.assignedPartitions()
    log.debug("Marking assigned partitions pending for revocation: {}", partitions);
    subscriptions.markPendingRevocation(partitions);
  }
}
```

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -935,6 +941,9 @@ private boolean isPaused() {
             return paused;
         }
 
+        private boolean isPendingRevocation() {
```
nit: do we need this? Is it used anywhere?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -853,12 +850,38 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
         return true;
     }
 
+    private SortedSet<TopicPartition> eagerPartitionsToRevoke(RebalanceProtocol protocol) {
```
nit: not using this anymore

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -853,12 +850,38 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
+        return partitions;
+    }
+
+    private void markPendingPartitions() {
```
How about `maybeMarkPartitionsPendingRevocation`? Otherwise it's a little unclear what exactly is pending.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java:
```diff
@@ -256,6 +256,15 @@ public void partitionPause() {
+        state.seek(tp0, 100);
+        assertTrue(state.isFetchable(tp0));
+        state.markPendingRevocation(singleton(tp0));
+        assertFalse(state.isFetchable(tp0));
```
Perhaps we can also assert `isPaused` is false?

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java:
```diff
@@ -2283,6 +2307,37 @@ public void testRestOffsetsAuthorizationFailure() {
     }
 
+    @Test
+    public void testPendingRevacationPartitionFetching() {
```
nit: Revocation is misspelled

I did not find the name very clear. It looks like the main difference between this and `testFetchingPendingPartitions` is that this method tests that the pending state gets reset after reassignment? Perhaps the name should reflect that?

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java:
```diff
@@ -272,6 +272,30 @@ public void testFetchNormal() {
+        assertEquals(4L, subscriptions.position(tp0).offset); // this is the next fetching position
+
+        // mark partition unfetchable
+        subscriptions.markPendingRevocation(singleton(tp0));
```
Another scenario is that we already have the fetch inflight when we mark pending revocation. Can we test that as well?

philipnee committed:
Test inflight fetches returns null.

hachikuji reviewed (approved):
Thanks, LGTM!

showuon commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java:
```diff
@@ -2283,6 +2322,37 @@ public void testRestOffsetsAuthorizationFailure() {
+    public void testFetchingPendingPartitionsBeforeAndAfterSubscriptionReset() {
+        buildFetcher();
+        assignFromUser(singleton(tp0));
+        subscriptions.seek(tp0, 100);
+        subscriptions.seek(tp0, 100);
+        subscriptions.seek(tp0, 100);
```
Any reason we seek tp0 to offset 100 three times?

showuon commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java:
```diff
@@ -2283,6 +2322,37 @@ public void testRestOffsetsAuthorizationFailure() {
+        subscriptions.seek(tp0, 100);
+        assertEquals(100, subscriptions.position(tp0).offset);
+
+        assertTrue(subscriptions.isFetchable(tp0)); // because tp is paused
```
I don't understand the comment here. Where do we pause tp0? and it is fetchable now, right?

philipnee replied:
I'll clean it up.

philipnee replied:
oops, this is a typo.  Probably some residual after modifying the tests.

showuon reviewed (approved):
Sorry for the delay review. LGTM! Left some minor comments, which can be addressed in a follow-up PR. Thank you.
